File: src/network.rs
use crate::cell::Cell;
use crate::content::{ State, Merge };
use crate::propagator::{Propagator, Procedure, Return};
use crate::util::{CellID, PropagatorID};
use crate::context::Context;
use std::fmt::Debug;

use std::collections::HashSet;

#[derive(Default)]
pub struct Network<C: Context, T> {
    cells: Vec<Cell<T>>,
    propagators: Vec<Propagator<C, T>>,

    context: C,

    alerted: HashSet<PropagatorID>
}

impl<C: Context, T> Network<C, T> {
    pub fn new(context: C) -> Self {
        Self {
            cells: Vec::new(),
            propagators: Vec::new(),
            
            context,

            alerted: HashSet::new()
        }
    }
}

impl<C: Context, T: Default> Network<C, T> {
    pub fn make_cell(&mut self) -> CellID {
        let cell = Cell::new();

        self.cells.push(cell);

        let id = self.cells.len() - 1;

        id
    }
}

impl<C: Context, T> Network<C, T> {
    pub fn label_cell(&mut self, id: CellID, label: &str) {
        self.cells[id].label = String::from(label);
    }

    pub fn label_propagator(&mut self, id: PropagatorID, label: &str) {
        self.propagators[id].label = String::from(label);
    }

    pub fn make_propagator(&mut self, proc: Procedure<C, T>, inputs: &[CellID], output: CellID) -> PropagatorID {
        let propagator = Propagator::new(proc, inputs, output);

        self.propagators.push(propagator);

        let id = self.propagators.len() - 1;

        for &cell_id in inputs {
            self.cells[cell_id].add_neighbour(id);
        }

        self.alerted.insert(id);

        id
    }

    pub fn alert_all_propagators(&mut self) {
        for id in 0..self.propagators.len() - 1 {
            self.alerted.insert(id);
        }
    }

    pub fn read_cell(&self, id: CellID) -> &T {
        self.cells[id].read()
    }
}

impl<C: Context, T: Debug + Merge + PartialEq> Network<C, T> {
    pub fn write_cell(&mut self, id: CellID, value: T) {
        let cell = &mut self.cells[id];
        let alerted = cell.write(value);

        if let Some(alerted) = alerted {
            for &propagator_id in alerted.iter() {
                self.alerted.insert(propagator_id);
            }
        }
    }
}

impl<C: Context, T> Network<C, T> 
where T: Debug + State + Clone + Merge + PartialEq {
    pub fn run(&mut self) {
        while self.alerted.len() > 0 {
            let mut writes : Vec<(CellID, Return<C, T>)> = Vec::new();

            for &propagator_id in self.alerted.iter() {
                let propagator = &self.propagators[propagator_id];

                let write_val = propagator.invoke(|&cell_id| {
                    self.cells[cell_id].read()
                });

                if let Some(write_val) = write_val {
                    writes.push(write_val);
                }
            }

            self.alerted.clear();

            writes.drain(0..).for_each(|(output_id, output)| {
                match output {
                    Return::Pure(output) => self.write_cell(output_id, output),
                    Return::Action(action) => {
                        self.context.run_action(action);

                        self.alert_all_propagators();
                    },
                }
            });
        }
    }
}

File: src/network.zig
const std = @import("std");
const warn = @import("std").debug.warn;
const ArrayList = std.ArrayList;
const Allocator = std.mem.Allocator;
const C = @import("cell.zig");
const P = @import("propagator.zig");
const CellID = @import("util.zig").CellID;
const PropagatorID = @import("util.zig").PropagatorID;

pub fn Network(comptime T: type, comptime Context: type) type {
    const Cell = C.Cell(T);
    const Propagator = P.Propagator(T);
    const Result = P.Result(T);

    return struct {
        const Self = @This();

        cells: ArrayList(Cell),
        propagators: ArrayList(Propagator),
        alerted: ArrayList(PropagatorID),
        context: *Context,
        allocator: *Allocator,

        pub fn init(allocator: *Allocator, context: *Context) Self {
            return Self {
              .cells = ArrayList(Cell).init(allocator),
              .propagators = ArrayList(Propagator).init(allocator),
              .alerted = ArrayList(PropagatorID).init(allocator),
              .context = context,
              .allocator = allocator,
            };
        }

        pub fn deinit(self: *Self) void {}

        pub fn make_cell(self: *Self) CellID {
          var cell = Cell.init(self.allocator);
          self.cells.append(cell) catch unreachable;
          
          return self.cells.items.len - 1;
        }

        pub fn write_cell(self: *Self, cell_id: CellID, content: *const T) void {
          var cell = self.cells.items[cell_id];

          const alerted = cell.write(content);

          self.cells.items[cell_id] = cell;
        
          if (alerted) |propagator_ids| {
              for (propagator_ids.items) |propagator_id| {
                self.alerted.append(propagator_id) catch unreachable;
              }
          }
        }

        pub fn read_cell(self: *Self, cell_id: CellID) T {
          var cell = self.cells.items[cell_id];

          return cell.content;
        }

        pub fn make_propagator(self: *Self, procedure: fn([]T) T, inputs: []CellID, output: CellID) PropagatorID {
            var propagator = Propagator.init(self.allocator, procedure, inputs, output) ;

            self.propagators.append(propagator) catch unreachable;

            const propagator_id = self.propagators.items.len - 1;

            for (inputs) |cell_id| {
                self.cells.items[cell_id].neighbours.append(propagator_id) catch unreachable; //FIXME
            }

            return propagator_id;
        }
      
        pub fn run(self: *Self) !void {
          while (self.alerted.items.len > 0) {
            var writes = try self.allocator.alloc(Result, self.alerted.items.len);
            var contents = try self.allocator.alloc(T, self.cells.items.len);
            var cur : usize = 0;
            //TODO defererr?

            for (self.cells.items) |*cell, i| {
              contents[i] = cell.content;
            }

            for (self.alerted.items[0..self.alerted.items.len]) |propagator_id| {
              var propagator = self.propagators.items[propagator_id];

              writes[cur] = try propagator.invoke(contents);
              cur += 1;
            }

            self.alerted = ArrayList(PropagatorID).init(self.allocator);

            for (writes) |value| {
              switch (value) {
                Result.Nothing => {},
                Result.Pure => |result| {
                  self.write_cell(result.cell_id, &result.content);
                },
              }
            }

            //TODO realloc?
            self.allocator.free(writes); 
            self.allocator.free(contents); 
          }
        }
    };
}


